File: crates/ask_pete_trainyard_web/src/types/NarrativeGraph.ts
export interface NarrativeGraph {
    nodes: Record<string, NarrativeNode>;
    start_node_id: string;
    metadata?: Record<string, string>;
}

export interface NarrativeNode {
    id: string;
    speaker: string;
    text: string;
    choices: NarrativeChoice[];
    events: NarrativeEvent[];
    literary_device?: string;
    quest?: Quest; // [NEW] Syncs Node with Quest System
    position?: NodePosition;
}

export interface Quest {
    title: string;
    chapter_theme: string;
    description: string;
    starting_step: string;
    completion_reward: QuestReward;
    steps: Record<string, QuestStep>;
}

export interface QuestReward {
    type: string;
    value?: number;
    details?: string;
    name?: string;
}

export interface QuestStep {
    description: string;
    choices?: NarrativeChoice[];
    trigger_condition: string;
    next_step?: string;
    step_reward?: QuestReward;
    is_major_plot_point?: boolean;
}

export interface NarrativeChoice {
    text: string;
    next_node_id?: string;
    conditions: NarrativeCondition[];
}

export interface NarrativeEvent {
    event_type: string;
    payload: Record<string, string>;
}

export interface NarrativeCondition {
    condition_type: string;
    parameters: Record<string, string>;
}

export interface NodePosition {
    x: number;
    y: number;
}

export type NodeData = Omit<NarrativeNode, 'id' | 'position'>;

// Backend Types (matching ask_pete_core::expert::StoryGraph)
export interface StoryGraph {
    id: string;
    title: string;
    nodes: BackendStoryNode[];
    connections: StoryConnection[];
}

export interface BackendStoryNode {
    id: string;
    title: string;
    content: string;
    x: number;
    y: number;
    passenger_count: number;
    complexity_level: number;
    learner_profiles: string[];
    gardens_active: string[];
    required_stats: Record<string, number>;
    logic: any; // LogicBlock
    style: any; // NodeStyle
    quest?: Quest;
}

export interface StoryConnection {
    id: string;
    from_node: string;
    to_node: string;
}

export interface BlueprintResponse {
    graph: StoryGraph;
    reasoning: string;
}

File: crates/ask_pete_core/src/narrative_graph.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NarrativeGraph {
    pub nodes: HashMap<String, NarrativeNode>,
    pub start_node_id: String,
    #[serde(default)]
    pub metadata: HashMap<String, String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NarrativeNode {
    pub id: String,
    pub speaker: String,
    pub text: String,
    #[serde(default)]
    pub choices: Vec<NarrativeChoice>,
    #[serde(default)]
    pub events: Vec<NarrativeEvent>,
    #[serde(default)]
    pub position: Option<NodePosition>, // For React Flow visualization
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NarrativeChoice {
    pub text: String,
    pub next_node_id: Option<String>,
    #[serde(default)]
    pub conditions: Vec<NarrativeCondition>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NarrativeEvent {
    pub event_type: String,
    pub payload: HashMap<String, String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NarrativeCondition {
    pub condition_type: String,
    pub parameters: HashMap<String, String>,
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct NodePosition {
    pub x: f32,
    pub y: f32,
}
